### PR TITLE
feat: expand network discovery to 4096 targets

### PR DESCRIPTION
**Pull request by ankitgoswami**
Reviewable diff: +99/-74 across 18 files (excludes generated, test, and story files).

## Summary

Fleet Node discovery accepts /20 subnets and up to 4,096 targets under one shared contract. The API and callers use the network_scan name, completing the Nmap replacement across the stack.

**Stack:** [#1035](https://github.com/block/proto-fleet/pull/1035) → [#1036](https://github.com/block/proto-fleet/pull/1036) → [#1037](https://github.com/block/proto-fleet/pull/1037) **(this PR)**

This diff is relative to its immediate base, [#1036](https://github.com/block/proto-fleet/pull/1036). [#1034](https://github.com/block/proto-fleet/pull/1034) supplies shared bounds and target semantics, [#1034](https://github.com/block/proto-fleet/pull/1034) and [#1035](https://github.com/block/proto-fleet/pull/1035) replace both runtimes, and [#1036](https://github.com/block/proto-fleet/pull/1036) preserves results through incomplete-source warnings.

This completes the planned implementation stack. Real miner, installed-package and full /20 deadline qualification remain deployment prerequisites; no merge or deployment is part of these drafts.

## How it works

The shared prefix minimum changes to /20 and derives the 4,096-target command cap. Protobuf list and raw-port limits align with execution, and mode field 4 is renamed to NetworkScanModeRequest/network_scan without a compatibility alias. The existing 1,024-report upload size and 64-event queue remain: at most forty report batches plus one ACK fit without evicting reports.

## Diagrams

```mermaid
flowchart LR
  R["network_scan, IP list or IP range"] --> V["Shared /20, 4096-target and 10-port limits"]
  V --> S["Server and node discovery implementations"]
  S --> B["Up to 40 batches of 1024 node reports"]
  B --> Q["Existing 64-event command queue"]
  A["Final ACK"] --> Q
  Q --> UI["Devices and incomplete-source warnings"]
```

## Areas of the code involved

| Area / package / file | What changed | Why it matters for review |
| --- | --- | --- |
| proto/pairing/v1 and fleetnodegateway/v1 | Rename mode field 4 and align discovery request limits | No new protocol version or compatibility aliases |
| Generated Go/TypeScript and Go/client callers | Regenerate and consume the renamed contract | Generated outputs — skip; all active callers move together |
| server/internal/domain/discoverylimits, Fleet Node upload and boundary tests | Raise the shared prefix/target cap, validate serialized commands and allow 90 seconds for upload | Dispatch limits agree with node transport; the total upload remains bounded |
| Fleet Node README, scripts and review-policy prompts | Describe TCP discovery and final bounds | Active operational and review guidance no longer depends on Nmap |

## Key technical decisions & trade-offs

**Why the scan limits change:** The previous /22 and 1,024-target bounds required operators to split larger networks into multiple requests. Allowing /20 subnets and 4,096 explicit targets lets one Fleet Node command cover a larger network while retaining a finite per-command workload. The shared Go target cap is derived from the prefix limit, so server dispatch validation and node execution stay aligned. A /20 contains 4,094 usable host addresses; explicit ranges can include all 4,096 addresses. Fleet Server retains its broader subnet policy and uncapped aggregation of known subnets.

The protobuf port limit drops from 256 to ten raw entries because the shared runtime already enforces ten: this makes API validation match execution rather than advertising unsupported input. Concurrency, the ten-minute scan deadline, and report size limits stay unchanged. The final upload budget increases from 30 to 90 seconds for all batches combined, leaving 30 seconds for dispatch and ACK delivery within the server’s existing 12-minute wait. At the new maximum, 4,096 addresses × ten ports produce at most forty 1,024-report batches plus an ACK, which fit the existing 64-event queue. This establishes reporting capacity, not a guarantee that every /20 finishes before the deadline; partial results and incomplete-source warnings remain necessary.

- Update components together; retain existing capability, protocol-version, upgrade and maintenance support.
- Keep ten raw ports and 1,024 reports per upload, with one bounded 90-second final upload; no queue redesign or progressive node upload protocol.
- Validate the serialized command with the node’s existing protobuf validator before dispatch. Oversized payloads produce a clear validation error/source warning asking the caller to split the request or shorten hostnames, without reducing ordinary 4,096-target scans.
- Keep server-known-subnet aggregates uncapped; only the node command and explicit list limits rise.
- Do not promise every /20 scan finishes in ten minutes: slow plugin responses can exhaust the command budget.

## Testing & validation

- Race tests cover /20 accepted versus /19 rejected, 4,096 versus 4,097 targets, ten versus eleven raw ports, and all forty report batches arriving before the ACK without eviction.
- Regression tests reject 4,096 distinct 253-byte hostnames before dispatch and accept the nearby fitting 252-byte case. A simulated-latency test uploads all forty batches over forty seconds under one shared deadline and verifies a stalled upload stops at ninety seconds; the forty-second case fails with the old budget. This is a deterministic timing regression, not a database throughput benchmark.
- A no-network cloud test scans 8,188 usable addresses from two /20 known subnets, proving aggregate expansion is not capped at 4,096.
- Full server/client lint, production server build, client type check, discovery UI tests, installer/profile checks and Windows amd64/Linux ARM64 node builds pass.
- Regenerated the affected protobuf output with the pinned protoc-gen-es 2.14.0 and repository Buf configuration. Protobuf lint, client type check and 33 focused hook/onboarding tests pass. A clean locked client installation was not rerun; hardware and database integration qualification were not completed locally.

### Post-deploy monitoring & validation

Release the complete stack with server, nodes and client updated together. On a controlled network, compare expected miners against both server and node discovery, exercise an incomplete scan, and confirm that retained results and its warning reach the client. Pause rollout if expected devices disappear, scans cannot stop within their budgets, or completed reports are lost. No deployment was performed for these draft PRs.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)

Restack validation: affected unit/race suites, server lint, protobuf lint, client type check and 33 focused client tests pass. Database integration tests remain for CI.

### CODE REVIEW

**Comment by ankitgoswami** [comment]
🤖

> The changed TypeScript protobuf files use protoc-gen-es 2.12.1 instead of the repository-pinned 2.14.0 generator.

Confirmed and not yet resolved. The local generator is 2.12.1; fetching pinned 2.14.0 is blocked by the company dependency-confusion policy at registry.npmjs.org. I did not bypass the block or hand-edit generated output. Regeneration needs an environment with the approved pinned toolchain. This is also tracked in #1036: https://github.com/block/proto-fleet/pull/1036#discussion_r3983655065.

**Comment by ankitgoswami** [comment]
🤖

> The changed TypeScript protobuf files use protoc-gen-es 2.12.1 instead of the repository-pinned 2.14.0 generator.

Registry access now works. Regenerated the affected protobuf output with the pinned protoc-gen-es 2.14.0 through the repository Buf configuration and pushed 398826816. Repeating generation and formatting leaves the working tree clean. Protobuf lint, TypeScript checking and all 33 focused client tests pass. This supersedes the earlier blocked-generation reply.

File: server/internal/domain/discoverylimits/limits.go [hunk 1]
```diff
@@ -5,7 +5,7 @@ package discoverylimits
 
 const (
 	// MinIPv4PrefixBits caps the breadth of an individual discovery subnet.
-	MinIPv4PrefixBits = 22
+	MinIPv4PrefixBits = 20
 
 	// MaxScanTargets caps IP addresses per Fleet Node discovery command.
 	MaxScanTargets = 1 << (32 - MinIPv4PrefixBits)
```
Comment (rl-block) [inline server/internal/domain/discoverylimits/limits.go:11]: Should this larger limit require a new command-protocol version or capability? Fleet Nodes from the current release still advertise V1 while enforcing /22 and 1,024 targets, so the new server will treat them as eligible and dispatch requests they reject as BAD_REQUEST. Perhaps V2/capability gating or server-side chunking could make rolling upgrades safe.

File: server/internal/domain/fleetnode/control/registry_test.go [hunk 2]
```diff
@@ -305,6 +305,45 @@ drain:
...
+	require.NoError(t, err)
+	defer session.Close()
+	require.Equal(t, "full-scan", recvCommandID(t, stream))
+
+	// A node can identify one device at every endpoint: 4,096 targets × 10 ports.
+	// End-only upload must fit even when the operator has not drained any events.
+	const reportBatchSize = 1024
+	require.Equal(t, 40960, maxReportsPerCommand)
+	const batches = 40
+	require.LessOrEqual(t, batches+1, commandEventBuffer)
```
Comment (rl-block) [inline server/internal/domain/fleetnode/control/registry_test.go:323]: This proves the events fit in memory, but the maximum path still performs 40,960 individual database upserts through 40 sequential RPCs under one 30-second upload deadline. It might be worth batching the persistence, streaming during the scan, or adding an end-to-end timing test so a valid maximum result does not end as REPORT_FAILED.

**Review by rl-block — APPROVED** [review]

File: server/internal/domain/discoverylimits/limits.go (hunk 1 above)
Comment (ankitgoswami) [reply server/internal/domain/discoverylimits/limits.go:11]: we only have one fleet-node right now, easier to just update it

File: proto/pairing/v1/pairing.proto [hunk 3]
```diff
@@ -94,28 +94,28 @@ message IPRangeModeRequest {
...
   // fully override the server-derived defaults.
-  repeated string ports = 3 [(buf.validate.field).repeated.max_items = 256];
+  repeated string ports = 3 [(buf.validate.field).repeated.max_items = 10];
 }
 
 // Configuration for discovering devices from a specific list of IP addresses
 message IPListModeRequest {
   // List of IP addresses (IPv4, IPv6, or hostnames) to check
-  repeated string ip_addresses = 1 [(buf.validate.field).repeated.max_items = 1024];
+  repeated string ip_addresses = 1 [(buf.validate.field).repeated.max_items = 4096];
```
[the remark this answers is not in the document]
Comment (ankitgoswami) [reply proto/pairing/v1/pairing.proto:103]: 🤖

> Keep 4,096-target requests within the command payload cap

Addressed in f6b7ac33e by validating the serialized ControlCommand with the same protobuf validator used by Fleet Node before dispatch. Oversized commands now return InvalidArgument with guidance to split the request or shorten hostnames. General discovery forwards this as a node warning while preserving server results; direct-node discovery returns the validation error.

This reuses the existing payload limit without adding an arbitrary hostname limit or a duplicate size constant. Regression tests reject 4,096 distinct valid 253-byte hostnames before dispatch and successfully dispatch the nearby fitting 252-byte case. Full lint and affected race suites pass.

File: server/internal/domain/fleetnode/control/registry_test.go (hunk 2 above)
Comment (ankitgoswami) [reply server/internal/domain/fleetnode/control/registry_test.go:323]: 🤖

> 40 sequential RPCs under one 30-second upload deadline

Addressed with a bounded budget increase in f6b7ac33e. All batches now share 90 seconds, leaving 30 seconds for dispatch and ACK delivery within the existing 12-minute server wait after a ten-minute scan. The budget is not reset per batch.

The regression uploads all 40,960 reports in forty batches with one second of simulated latency per RPC. It completes over forty simulated seconds with the new budget and fails with the old thirty-second budget. A stalled upload still stops at the total deadline. Full lint and affected race suites pass.

This provides upload headroom without changing persistence or introducing progressive uploads. The timing test simulates latency; it is not a real-database benchmark, and slow deployments may still return partial results. The PR description and node documentation state that limitation.
